kyaml/fn/framework/framework.go

```diff
@@ -48,14 +54,159 @@ type ResourceList struct {
 	//      kind: Example
 	//      spec:
 	//        foo: var
-	FunctionConfig *yaml.RNode `yaml:"functionConfig,omitempty" json:"functionConfig,omitempty"`
+	FunctionConfig *yaml.KubeObject `yaml:"functionConfig,omitempty" json:"functionConfig,omitempty"`
```

Comment:
I played with this branch a bunch yesterday, and I will comment in more detail later. But I wanted to start by sharing the central challenge I encountered in attempting to use this branch for a non-trivial use case, which comes down to the changes to these fields. In addition to the compatibility concerns @yuwenma raised (we can making breaking changes since this is alpha, but we should still be somewhat conservative about it for existing consumers' sake), not having these be RNodes breaks compatibility with the existing suite of tools. Everything from Processors to Matchers/Selectors revolves around the `kyaml.Filter` interface, which has a simple "rnodes in, rnodes out" contract that makes it really powerful. I'm in favour of making simple use cases easier by layering on top of this, but I don't think we should move away from RNode and Filter being the primitive concepts in the foundational layers (e.g. in the type definition here). 

Related to the above, I continue to be skeptical that it will be a net benefit to introduce `KubeObject` at the same level of abstraction as `RNode` on the basis that the existing method surface of `RNode` is too big. Having two types represent the same thing but work with different parts of the package could be more complexity overall, as it is something consumers need to learn to decide (and convert!) between to use our tools. For example, in a method below, you have to re-invent the signature for selector as `selector func(obj *yaml.KubeObject) bool`, which is then incompatible with the large suite of high-level selectors and low-level selector builders that the framework already has (see `selector.go` and `matchers.go`).

Reply:
I'm OK with merging `KubeObject` with `RNode` for now.

IMO when users uses `kyaml/fn/framework` package, they don't need to deal with io directly. IO should be handled by the framework package.
Also `kio.Filter` is doing similar things as `ResourceListProcessor`. This is another confusing thing for users. IMO the framework package should move to `ResourceListProcessor` (`ResourceList` in, `ResourceList` out) eventually, since `ResourceList` is in the contract of KRM function spec and it allows structured results to be passed.